File: system-test/dotnet/Driver.Adapter/External/Clock/ClockStubDriver.cs
using Driver.Port.External.Clock;
using Driver.Port.External.Clock.Dtos;
using Driver.Adapter.External.Clock.Client;
using Driver.Adapter.External.Clock.Client.Dtos;
using Driver.Adapter.External.Clock.Client.Dtos.Error;
using Common;

namespace Driver.Adapter.External.Clock;

public class ClockStubDriver : IClockDriver
{
    private readonly ClockStubClient _client;
    private bool _disposed;

    public ClockStubDriver(string baseUrl)
    {
        _client = new ClockStubClient(baseUrl);
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (_disposed) return;
        if (disposing)
            _client?.Dispose();
        _disposed = true;
    }

    public Task<Result<VoidValue, ClockErrorResponse>> GoToClockAsync()
        => _client.CheckHealthAsync().MapErrorAsync(MapError);

    public Task<Result<GetTimeResponse, ClockErrorResponse>> GetTimeAsync()
        => _client.GetTimeAsync().MapAsync(MapResponse).MapErrorAsync(MapError);

    public async Task<Result<VoidValue, ClockErrorResponse>> ReturnsTimeAsync(ReturnsTimeRequest request)
    {
        var extResponse = new ExtGetTimeResponse
        {
            Time = DateTimeOffset.Parse(request.Time!, System.Globalization.CultureInfo.InvariantCulture)
        };
        await _client.ConfigureGetTimeAsync(extResponse);
        return Result.Success<ClockErrorResponse>();
    }

    private static GetTimeResponse MapResponse(ExtGetTimeResponse response)
        => new GetTimeResponse { Time = response.Time };

    private static ClockErrorResponse MapError(ExtClockErrorResponse errorResponse)
        => new ClockErrorResponse { Message = errorResponse.Message };
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/external/clock/ClockStubDriver.java
package com.mycompany.myshop.testkit.driver.adapter.external.clock;

import com.mycompany.myshop.testkit.driver.port.external.clock.ClockDriver;

import com.mycompany.myshop.testkit.driver.adapter.external.clock.client.ClockStubClient;
import com.mycompany.myshop.testkit.driver.adapter.external.clock.client.dtos.ExtGetTimeResponse;
import com.mycompany.myshop.testkit.driver.port.external.clock.dtos.GetTimeResponse;
import com.mycompany.myshop.testkit.driver.port.external.clock.dtos.ReturnsTimeRequest;
import com.mycompany.myshop.testkit.driver.port.external.clock.dtos.error.ClockErrorResponse;
import com.mycompany.myshop.testkit.common.Closer;
import com.mycompany.myshop.testkit.common.Converter;
import com.mycompany.myshop.testkit.common.Result;

public class ClockStubDriver implements ClockDriver {
    private final ClockStubClient client;

    public ClockStubDriver(String baseUrl) {
        this.client = new ClockStubClient(baseUrl);
    }

    @Override
    public void close() {
        client.removeStubs();
        Closer.close(client);
    }

    @Override
    public Result<Void, ClockErrorResponse> goToClock() {
        return client.checkHealth()
                .mapError(ext -> new ClockErrorResponse(ext.getMessage()));
    }

    @Override
    public Result<GetTimeResponse, ClockErrorResponse> getTime() {
        return client.getTime()
                .map(ext -> GetTimeResponse.builder().time(ext.getTime()).build())
                .mapError(ext -> new ClockErrorResponse(ext.getMessage()));
    }

    @Override
    public Result<Void, ClockErrorResponse> returnsTime(ReturnsTimeRequest request) {
        var time = Converter.toInstant(request.getTime());

        var extResponse = ExtGetTimeResponse.builder()
                .time(time)
                .build();

        return client.configureGetTime(extResponse)
                .mapError(ext -> new ClockErrorResponse(ext.getMessage()));
    }
}
